File: cpp_recursive/Cluster.cpp
#include "Cluster.h"
#include <bits/stdc++.h>
#include "DF_Headers.h"
#include "Prints.h"
using namespace std;

// default constructor
// needed for default creation in builds map, but there are branches in functions (that return 0) just in case
Building::Building() {
	info = nullptr;
	name = "";
	maxCPS = 0;
	multiplier = 0;
	quantity = 1;
	// workers.clear()
	// fullCPS = false;
}

// input is address of the dataframe row and quantity
Building::Building(vector<string>* building, int num = 1) : info(building) {
	name = info->at(Name);
	maxCPS = stoi(info->at(MaxCPS));
	multiplier = stoi(info->at(Multiplier).substr(1));
	quantity = num;

	for (int i = Worker1; i <= Worker6 && !building->at(i).empty(); i++) {
		workers[building->at(i)] = 0;
	}
}

// print workers
// TODO: add CPS stuff
string Building::str() const {
	string ans = name;
	
	ans += " (" + to_string(this->maxCPS) + ") " + to_string(this->getCPS());
	ans += " : {";
	for (map<string, int>::const_iterator i = workers.cbegin(); i != workers.cend(); i++) {
		if (i != workers.cbegin())
			ans += ", ";
		ans += i->first + ": " + to_string(i->second);
	} ans += "}";
	return ans;
}

// true if job can and was added, false if not
bool Building::insertJob(const string& j) {
	if (info == nullptr)
		return false;

	if (workers.find(j) != workers.end() && workers[j] < quantity) {
		workers[j]++;
		return true;
	}

	return false;
}
// true if job exists and was removed, false if not
// go from the back of the list to make cps front-heavy
bool Building::removeJob(const string& j) {
	if (info == nullptr)
		return false;

	if (workers.find(j) != workers.end() && workers[j] > 0) {
		workers[j]--;
		return true;
	}
	return false;
}

// calculate all building CPS based on number of workers
int Building::getCPS() const {
	if (info == nullptr)
		return 0;
	
	int min = workers.begin()->second, total = 0;
	for (map<string, int>::const_iterator i = workers.cbegin(); i != workers.cend(); i++) {
		total += i->second;
		if (i->second < min)
			min = i->second;
	}

	return (total + min*workers.size()) * multiplier;
}

// -------------------

Cluster::Cluster(map<string, vector<string>> *buildings, map<string, vector<string>> *jobs, const map<string, int>& qbmap, const map<string, int>& qpmap) :
buildings(buildings), jobs(jobs), qbmap(qbmap), qpmap(qpmap) {
	maxVal = 0;
	done = false;
}

// print max and config
string Cluster::str() {
	if (!done) {
		getMax();
	}

	string ans = "Max: ";
	ans.append(to_string(getMax()));
	ans.push_back('\n');
	ans.append(configStr());
	ans.push_back('\n');

	return ans;
}

// assign the jobs that only have one workplace
// make changes to parameters
void Cluster::filterEasy(map<string, Building>& builds, vector<string>& people) {
	vector<string> newp;
	string currPerson;
	vector<string> currRow;
	for (unsigned i = 0; i < people.size(); i++) {
		currPerson = people.at(i);
		currRow = jobs->at(currPerson);
		cout << currPerson << endl;
		
		// needs to have only 1 workplace.
		// should have a spot guaranteed
		if (!(stoi(currRow.at(Options)) == 1 && builds[currRow.at(Workplace1)].insertJob(currPerson))) {
			newp.push_back(people.at(i));
		}
	}
	people = newp;
}
// assign jobs where we have a one-to-one quantity between the job and their buildings
// 	remove curr later
void Cluster::filterFit(map<string, Building>& builds, map<string,int>& qpmap) {
	int sum;
	vector<int>curr(3);
	vector<string> currRow;
	for (auto i: qpmap) {
		sum = 0;
		curr.resize(3,0);
		currRow = jobs->at(i.first);
		// get number of buildings
		for (unsigned j = 0; j < 3 && !currRow.at(j).empty(); j++) {
			if (builds.find(currRow.at(j+Workplace1)) != builds.end()) {
				curr.at(j) += builds[currRow.at(j+Workplace1)].quantity;
				sum += curr.at(j);
			}
		}

		// if the numbers match, put them in buildings and remove the job
		if (sum == i.second) {
			for (unsigned j = 0; j < 3 && !currRow.at(j).empty(); j++) {
				while (curr.at(j) > 0) {
					builds[currRow.at(j+Workplace1)].insertJob(i.first);
					curr.at(j)--;
				}
			}
			qpmap.erase(i.first);
		}
		curr.clear();
	}
}
// iterate through all of the buildings and get their CPS's
int Cluster::calcCPS(const map<string, Building>& bdict) const {
	int sum = 0;
	for (auto& i: bdict) {
		sum += i.second.getCPS();
	}
	return sum;
}
void printState(map<string, Building>& currBuilds) {
	cout << "State:\n{";
	for (map<string, Building>::iterator i = currBuilds.begin(); i != currBuilds.end(); i++) {
		if (i != currBuilds.begin()) {
			cout << ", ";
		}
		cout << i->first << " : " << i->second.str();
	}
	cout << "}\n";
}

struct cmp {
	bool operator()(const pair<int, string>& p1, const pair<int, string>& p2) {
		return p1.first > p2.first;
	}
};
void Cluster::recursion(map<string, Building>& currBuilds, const vector<string>& plist, unsigned int& count, unsigned int iter = 0) {
	if (plist.empty()) {
		// cout << "empty" << endl;
		return;
	}

	if (iter == plist.size()) {
		int currCPS = calcCPS(currBuilds);
		if (currCPS > maxVal) {
			maxVal = currCPS;
			maxMap = currBuilds;
		}

		count += 1;
		if (count % 100000 == 0) {
			cout << fixed << setprecision(3) << time(0)-startTime << " - " << setprecision(0) << count << endl;
		}
		return;
	}

	// for each job, try to assign it to a workplace
	// if it works, recurse through to next job
	// else, skip
	for (int i = Workplace1; i <= Workplace3; i++) {
		string build = jobs->at(plist[iter]).at(Workplace1 + i);
		if (!build.empty()) {
			currBuilds[build].insertJob(plist[iter]);
		}

		recursion(currBuilds, plist, count, iter+1);

		currBuilds[build].removeJob(plist[iter]);
	}
}
int Cluster::getMax() {
	if (done) {
		return maxVal;
	}

	// set up data
	map<string, Building> builds;
	vector<string> *row;

	for (auto b : this->qbmap) {
		row = &buildings->at(b.first);
		Building temp(row);
		builds[b.first] = temp;
	}

	map<string, int> qpmap = this->qpmap;
	filterFit(builds, qpmap);		// useful to make obvious choices

	vector<string> people;
	for (map<string, int>::iterator i = qpmap.begin(); i != qpmap.end(); i++) {
		while (i->second > 0) {
			people.push_back(i->first);
			qpmap[i->first]--;
		}
	}

	filterEasy(builds, people);		// useful to make obvious choices
	maxVal = calcCPS(builds);
	maxMap = builds;
	cout << "Trimmed. " << people.size() << " people remain." << endl;

	cout << "About 3^" << people.size() << " permutations." << endl;

	startTime = time(0);
	unsigned count = 0;			// used for update log
	recursion(builds, people, count);
	cout << "Finished calculation.\n";
	cout << "Total time: " << time(0)-startTime << "\n";
	cout << endl;

	done = true;

	return maxVal;
}

string Cluster::configStr() {
	if (!done) {
		getMax();
	}

	string ans;
	// can't use printMap(), so gotta write it normally
	ans = "";
	for (map<string, Building>::iterator i = maxMap.begin(); i != maxMap.end(); i++) {
		if (i != maxMap.begin())
			ans.append("\n");
		ans.append(i->second.str());
	}

	return ans;
}


// create a vector of sets that have all connected buildings and people
vector<set<string>> generateClusters(const map<string, vector<string>>& buildings) {
	vector<set<string>> ans;

	// for each building
	// get the jobs that can go there and throw them all in a set
	// if we encounter a job that we have already seen, union the sets
	// finally, push to ans
	set<string> temp;
	for (auto& i: buildings) {						// for each line of buildings
		temp.clear();

		temp.insert(i.second.at(Building_name));
		for (unsigned j = Worker1; j <= Worker6; j++) {
			if (!i.second[j].empty()) {
				temp.insert(i.second[j]);
			}
		}
		// cout << setToStr(temp) << endl;

		for (unsigned k = 0; k < ans.size(); k++) {		// for every set saved
			for (auto& j: temp) {						// iterate through temp, looking for values
				if (ans.at(k).find(j) != ans.at(k).end()) {
					temp.merge(ans.at(k));
					ans.at(k).clear();
					break;
				}
			}
		}
		ans.push_back(temp);
	}

	// clean up ans by removing all empty sets used
	vector<set<string>> ret;
	for (unsigned i = 0; i < ans.size(); i++) {
		if (!ans.at(i).empty())
			ret.push_back(ans.at(i));
	}

	// sort vector by set size
	auto cmp = [](const set<string>& s1, const set<string>& s2) { return s1.size() < s2.size(); };
	sort(ret.begin(), ret.end(), cmp);

	cout << ret.size() << " clusters generated.\n";
	
	return ret;
}

File: archive/python_recursive/Cluster.py
import copy
import pandas as pd
import time

def generateClusters(buildings):		# jobs df not needed
	# filter through data to get clusters to work on individually rather than as a whole
	# just do it on the whole thing
	sets = []

	for b in buildings.index:
		curr = {b}
		for p in buildings.loc[b]['Worker1':'Worker6']:
			if not pd.isna(p):
				curr.add(p)
		existsIn = []
		for i in range(len(sets)):
			for s in curr:
				if s in sets[i] and i not in existsIn:
					existsIn.append(i)					# duplicate indices are NOT fine
		
		print(len(sets), existsIn)
		while len(existsIn) > 0:
			curr = curr.union(sets[existsIn[-1]])				# must be added/removed from the back to keep sets[] indices intact
			sets.remove(sets[existsIn[-1]])
			existsIn.pop()
		sets.append(curr)

	sets = sorted(sets, key=lambda x: len(x))
	# print(sets)
	return sets

# state = [int, {'Building': Building()}]
def printState(curr):
	print('\nState:')
	max = 0
	if type(curr[0]) is int:
		max = 1
		print(curr[0])

	for key in curr[0+max]:
		print(key, curr[0+max][key],end=', ')
	# print(curr[1+max])

# can do individual clusters of permutations
class Cluster:
	def __init__(self, buildings,jobs,qbmap,qpmap):
		self.buildings = buildings
		self.jobs = jobs
		self.qbmap = qbmap
		self.qpmap = qpmap
		
		self.max = [0, {}]
		self.done = False

		# print(self.qbmap, self.qpmap)
	def __str__(self):
		if not self.done:
			print("Warning: self.getMax() not completed.")
		# print value
		ret = 'Val ' + str(self.max[0]) + '\nBuildings: {'
		
		# print buildings
		buildings = copy.deepcopy(self.max[1])
		reversed(buildings)
		for key in buildings:
			ret += key + ' : ' + str(buildings[key]) + ', '
		ret = ret[:-2] + '}\n'
		
		# if len(self.max[2]) > 0:
		# 	ret += 'People: {'

		# 	# print people
		# 	people = copy.deepcopy(self.max[2])
		# 	for key in people:
		# 		ret += key + ' : ' + people[key] + ', '
		# 	ret = ret[:-2] + '}\n'
		
		return f"{ret}"

	class Building:
		# input: number of buildings
		def __init__(self, name, buildings, num=1):
			# self.buildings = buildings
			# self.jobs = [set() for i in range(int(num))]
			self.jobs = {}
			for b in buildings.loc[name]['Worker1':'Worker6']:
				if pd.notna(b):
					self.jobs[b] = 0
			self.mult = buildings.loc[name]['Multiplier']
			self.maxCPS = buildings.loc[name]['MaxCPS']			# for printing
			self.quantity = num

		def __str__(self):
			return str(self.getCPS()) + ' (' + str(self.maxCPS) + '): ' + str(self.jobs)
		def __len__(self):
			return len(self.jobs)
		

		# input: job that we want to assign
		# returns True if success, False if fail
		# no need to error check since jobs will be selected by those whose workplace is here
		def insertJob(self, j: str) -> bool:
			lastCPS = self.getCPS
			if j in self.jobs and self.jobs[j] < self.quantity:
				self.jobs[j] += 1
				return True
			return False
		
		# input: job that we want to assign
		# returns True if success, False if fail
		def removeJob(self, j: str) -> bool:
			lastCPS = self.getCPS
			if j in self.jobs and self.jobs[j] > 0:
				self.jobs[j] -= 1
				return True
			return 0
		
		def getCPS(self) -> int:
			# print(name, i)
			# print(buildings.loc[name])

			full = min(self.jobs.values()) * len(self.jobs.keys())
			sumVal = sum(self.jobs.values())
			ans = sumVal + full

			return ans * self.mult

	# do the jobs that only have one workplace first, or if we have a one-to-one quantity
	# will automatically update self.max
	def filterEasy(self, builds, people):
		# preload the buildings with the no-brainer options
		# append the no-brainers to the phash at the end
		ez_people = {}		# dict for all no-brainers
		poplist = []
		for i in people:
			if int(self.jobs.loc[i[:-1]]['Options']) == 1:
				ez_people[i] = self.jobs.loc[i[:-1]]['Workplace1']
				# self.max[2][i] = ez_people[i]
				
				builds[ez_people[i]].insertJob(i[:-1])
				self.max[1][ez_people[i]].insertJob(i[:-1])
				poplist.append(i)
				
		for i in poplist:
			# people.pop(i)
			del people[i]
		return ez_people
	
	# if qperson == qbuildings they can be assigned, do those
	def filterFit(self, builds, qpmap):
		# count = []
		poplist = []
		for i in qpmap:
			count = []
			workplaces = self.jobs.loc[i]['Workplace1':'Workplace3'].dropna().to_list()
			for w in workplaces:					# get the number of buildings (quantity) that this person can go to
				count.append(len(builds[w])) if w in builds else count.append(0)
			if sum(count) == qpmap[i]:
				for c in range(len(count)):
					if workplaces[c] in builds:
						while count[c] > 0:
							builds[workplaces[c]].insertJob(i)
							self.max[1][workplaces[c]].insertJob(i)
							count[c] -= 1
							qpmap[i] -= 1
				poplist.append(i)

		for i in poplist:
			# qpmap.pop(i)
			del qpmap[i]
		# return ez_people

	def calcCPS(self, bdict) -> int:
		sum = 0
		for i in bdict:
			sum += bdict[i].getCPS()
		return sum

	# iterate through keys and get a value to save
	def recursion(self, blist, plist, curr, iter=0, count=[0,0]):
		# print('Function begin:')
		# base case: we have reached the end of the people
		if iter == len(plist):
			val = self.calcCPS(curr[0])
			if val > self.max[0]:
				self.max[0] = val
				self.max[1] = copy.deepcopy(curr[0])
				# self.max[2] = dict(curr[1])

			count[0] += 1
			curr_time = time.time() - self.start_time
			if count[0] % 10000 == 0 or count[0] == count[1]:
				print("{:.3f}".format(curr_time), 'Generated', count[0], 'of', count[1], 'permutations.', 'Current CPS:', self.max[0])
				if count[0] > 0 and count[0] % 100000 == 0:
					print(self)
				
			return
		
		# iterative step: set one of the jobs to a building, then recursively iterate through the rest
		
		workplaces = self.jobs.loc[plist[iter][:-1]]['Workplace1':'Workplace3'].dropna().to_list()
		# workplaces = [''] + workplaces				# assume every person will not have a place to go

		for w in workplaces:
			if not pd.isna(w):
			
				if w in blist and curr[0][w].insertJob(plist[iter][:-1]):			# add job to building. make sure assignment worked
					curr[1][plist[iter]] = w
					pass

				# printState(curr)
				
				# go to next level
				self.recursion(blist, plist, curr, iter+1, count)

				if w in blist:
					curr[0][w].removeJob(plist[iter][:-1])
					curr[1][plist[iter]] = ''


	# input: building_data, job_data, building_quantity, people_quantity
	def getMax(self):
		if self.done:
			return self.max
		
		# set up data
		builds = {}
		for b in self.qbmap.keys():
			temp = self.Building(b, self.buildings, 1)		# works for 1 building. need to add more buildings later
			builds[b] = temp
		self.max[1] = copy.deepcopy(builds)

		qpmap = dict(self.qpmap)
		self.filterFit(builds, qpmap)

		people = {}
		for p in qpmap.keys():
			while qpmap[p] > 0:
				people[p+str(int(qpmap[p]))] = ''
				qpmap[p] -= 1

		ez_people = self.filterEasy(builds, people)		# filter the easy ones out

		curr = [builds, people]
		print('new size:', len(builds), len(people))

		total = 1
		for i in people:
			num = float(self.jobs.loc[i[:-1]]['Options'])
			# num += 1.0				# +1 for empty
			total *= num
		print('Need to generate',total,'cases.')
		count = [0, total]				# show a load time

		# put it into recursion
		self.start_time = time.time()
		self.recursion(tuple(builds.keys()), tuple(people.keys()), curr, count=count)
		print('Finished recursion')
		print('Total time:', time.time()-self.start_time)

		# for i in ez_people:
		# 	self.max[2][i] = ez_people[i]

		# printState(self.max)
		self.done = True
		# input('.')

		return self.max